test/test_overrides.py

```diff
@@ -0,0 +1,756 @@
+import torch
+import numpy as np
+import unittest
+import six
+import inspect
+import pprint
+
+from common_utils import TestCase
+
+HANDLED_FUNCTIONS = {}
+HANDLED_FUNCTIONS_SUB = {}
+HANDLED_FUNCTIONS_SUB_DIAGONAL = {}
+HANDLED_FUNCTIONS_TENSOR_LIKE = {}
+
+def implements(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS[torch_function.__name__] = func
+        return func
+    return decorator
+
+def implements_sub(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS_SUB[torch_function.__name__] = func
+        return func
+    return decorator
+
+def implements_sub_diagonal(torch_function):
+    "Register an implementation of a torch function for a Tensor-like object."
+    def decorator(func):
+        HANDLED_FUNCTIONS_SUB_DIAGONAL[torch_function.__name__] = func
```

Comment:
There may be conflicting names in different submodules in the future.

Reply:
Yes, I pointed out to @rgommers and @prasunanand a few days ago that using the names here instead of the function objects might be problematic. Do one of you two have a better understanding why the current implementation uses the names of the functions instead of the function objects?

Reply:
Yeah, we  couldn't pass function object as key initially from C++, we used names instead.

Reply:
See `PyUnicode_FromString(r.get_func_name().data())` on  [line](https://github.com/pytorch/pytorch/pull/27064/files#diff-64c8380f1aafb327ad41c479f79858cfR122) .

Reply:
I am also in favor of passing function object instead of function name as parameter to `PyObject_CallFunctionObjArgs()` .

Reply:
How about using the module plus the name? Like `torch.add`?

Reply:
Agree. This can be easily done.

Reply:
Then you can also get the function from the module and qualified name. 😄

Reply:
I have the opposite question: how does Numpy do keying based on functions rather than strings?

Reply:
It generates a wrapper based on the main function, and passes the wrapper into `__array_function__` inside the wrapper itself. Kinda like recursion but not really, since you're only passing in the function, not necessarily calling it.